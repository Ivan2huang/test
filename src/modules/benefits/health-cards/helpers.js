/* eslint-disable import/prefer-default-export */
export const getQRCodeXML = ({
  fullName,
  membershipNumber,
  certificateNumber,
  insurerCode,
  insurerName,
  insurerType,
  policyNumber,
  expiryDate,
}) =>
  [
    '<?xml version="1.0"?>',
    '<medical_card>',
    '<Card_type>',
    insurerType,
    '</Card_type>',
    '<insurance_company_code>',
    insurerCode,
    '</insurance_company_code>',
    '<insurance_company_name>',
    insurerName,
    '</insurance_company_name>',
    '<insured_name>',
    fullName,
    '</insured_name>',
    '<policy_no>',
    policyNumber,
    '</policy_no>',
    '<certificate_no>',
    certificateNumber,
    '</certificate_no>',
    '<membership_no>',
    membershipNumber,
    '</membership_no>',
    '<expiry_date>',
    expiryDate,
    '</expiry_date>',
    '</medical_card>',
  ].join('');
