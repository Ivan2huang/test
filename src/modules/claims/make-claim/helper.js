import moment from 'moment';
import {
  formatAmount,
  formatMessage,
  formatDate,
} from '../../../helpers/helpers';
import { DATE_WITHOUT_ZONE } from '../constant';

export const patientDetails = (patientName, contactNumber, intl) => [
  {
    label: formatMessage(
      intl,
      'claims.claimDetails.label.patientName',
      'Patient name',
    ),
    value: patientName,
  },
  {
    label: formatMessage(
      intl,
      'claims.claimDetails.label.contactNumber',
      'Contact number',
    ),
    value: contactNumber || '-',
  },
];

export const claimDetails = (
  {
    claim: {
      anotherInsurer,
      consultationDate,
      receiptAmount,
      isMaternity,
      otherInsurerAmount,
      diagnosisText,
    },
    consultationType,
    selectedClaimReason,
  },
  intl,
) => {
  const diagnosis = [selectedClaimReason, diagnosisText]
    .filter(Boolean)
    .join(' - ');
  return [
    {
      label: formatMessage(
        intl,
        'claims.claimDetails.label.consultationDate',
        'Consultation date',
      ),
      value: formatDate(consultationDate),
    },
    {
      label: formatMessage(
        intl,
        'claims.claimDetails.label.consultationType',
        'Consultation type',
      ),
      value: consultationType,
    },
    {
      label: formatMessage(
        intl,
        'claims.claimDetails.label.diagnosis',
        'Diagnosis',
      ),
      value: diagnosis,
    },
    {
      label: formatMessage(
        intl,
        'claims.claimDetails.label.receiptAmount',
        'Receipt amount',
      ),
      value: `HK$ ${formatAmount(intl, receiptAmount)}`,
    },
    {
      label: formatMessage(
        intl,
        'claims.claimReview.isMaternity.label',
        'Related to maternity?',
      ),
      value: isMaternity
        ? formatMessage(intl, 'claims.claimReview.isMaternity.yes', 'Yes')
        : formatMessage(intl, 'claims.claimReview.isMaternity.no', 'No'),
    },
    (anotherInsurer && otherInsurerAmount) > 0 && {
      label: formatMessage(
        intl,
        'claims.claimDetails.label.claimAmountFromAnotherInsurer',
        'Claim amount (other insurer)',
      ),
      value:
        otherInsurerAmount > 0 &&
        `HK$ ${formatAmount(intl, otherInsurerAmount)}`,
    },
  ].filter(Boolean);
};

export const transformConsultationDate = date =>
  moment(date, moment.ISO_8601)
    .startOf('day')
    .format(DATE_WITHOUT_ZONE);
