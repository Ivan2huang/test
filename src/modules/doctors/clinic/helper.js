// eslint-disable-next-line import/prefer-default-export
import {
  getValueFromLocalStorage,
  saveToLocalStorage,
} from '../../../helpers/helpers';
import CONFIG from '../../../constants/config';

const RECENT_SEARCH_KEY = 'clinic-recent-searches';

const isIncludes = checker => value =>
  value
    .toLowerCase()
    .trim()
    .includes(checker);

export const findClinics = (clinics, searchKeyword, searchBy = '') => {
  if (searchKeyword) {
    const searchKeywordLowerCase = searchKeyword.toLowerCase();
    if (searchBy) {
      return clinics.filter(
        clinic =>
          clinic[searchBy].toLowerCase().trim() === searchKeywordLowerCase,
      );
    }

    const isIncludesSearchValue = isIncludes(searchKeywordLowerCase);
    return clinics.filter(clinic => {
      const { name, area, district, address } = clinic;
      const hasName = isIncludesSearchValue(name);
      const hasArea = isIncludesSearchValue(area);
      const hasDistrict = isIncludesSearchValue(district);
      const hasAddress = isIncludesSearchValue(address);

      return hasName || hasArea || hasDistrict || hasAddress;
    });
  }
  return clinics;
};

export const filterClinics = (clinics, filters, filterBy = '') => {
  const filtersList = Object.keys(filters).filter(element => filters[element]);
  if (filtersList.length <= 0) {
    return clinics;
  }
  return clinics.filter(clinic => filters[clinic[filterBy]]);
};

export const getRecentSearch = () => {
  return JSON.parse(getValueFromLocalStorage(RECENT_SEARCH_KEY)) || [];
};

export const saveRecentSearch = (keyword, searchBy = '') => {
  const recentSearches = getRecentSearch();
  const indexOfDuplicate = recentSearches.findIndex(
    search => search.keyword === keyword && search.searchBy === searchBy,
  );
  if (indexOfDuplicate !== -1) {
    recentSearches.splice(indexOfDuplicate, 1);
  }
  if (recentSearches.length === CONFIG.maxRecentSearch) {
    recentSearches.pop();
  }
  recentSearches.unshift({ keyword, searchBy });
  saveToLocalStorage(RECENT_SEARCH_KEY, JSON.stringify(recentSearches));
};
