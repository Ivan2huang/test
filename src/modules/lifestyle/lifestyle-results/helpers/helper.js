import createBmiCardInfo from './createBmiCardInfo';
import createDiabetesInfo from './createDiabetesInfo';
import createAlcoholRiskInfo from './createAlcoholRiskInfo';
import createTobaccoRiskInfo from './createTobaccoRiskInfo';
import createExerciseRiskInfo from './createExerciseRiskInfo';
import createNutritionRiskInfo from './createNutritionRiskInfo';
import createSleepRiskInfo from './createSleepRiskInfo';
import createMindAndStressRiskInfo from './createMindAndStressRiskInfo';

const getResultCards = (intl, theme, lifestyleResults) => {
  if (Object.entries(lifestyleResults).length === 0) return [];
  return [
    createBmiCardInfo(
      intl,
      theme,
      lifestyleResults.bmiScore,
      lifestyleResults.bmiClass,
    ),
    createDiabetesInfo(intl, theme, lifestyleResults.diabetesRisk),
    createAlcoholRiskInfo(intl, theme, lifestyleResults.alcoholRisk),
    createTobaccoRiskInfo(intl, theme, lifestyleResults.tobaccoRisk),
    createExerciseRiskInfo(intl, theme, lifestyleResults.exerciseRisk),
    createNutritionRiskInfo(intl, theme, lifestyleResults.nutritionRisk),
    createSleepRiskInfo(intl, theme, lifestyleResults.sleepRisk),
    createMindAndStressRiskInfo(
      intl,
      theme,
      lifestyleResults.mindAndStressRisk,
    ),
  ];
};

export default getResultCards;
