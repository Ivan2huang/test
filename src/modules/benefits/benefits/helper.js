import { formatMessage } from '../../../helpers/helpers';
import { TERMINATED } from '../constant';

const WELLNESS_FLEXIBLE_SPENDING = 'WellnessFlexibleSpending';

const getDisplayName = (intl, name, plan) => {
  if (!plan) return name;
  const planName = plan.name;
  const defaultMessage = `${name} ( Tier ${planName} - Employee and dependant)`;
  return formatMessage(intl, 'me.tabs.benefits.dropdown.item', defaultMessage, {
    name,
    planName,
  });
};

export const buildMap = (data, groupBy) => {
  const map = {};
  data.forEach(item => {
    const key = item[groupBy];
    map[key] = map[key] ? map[key] + 1 : 1;
  });
  return map;
};

export const buildBenefitsSummaryList = (intl, memberProfile) => {
  const summaryList = [];
  const {
    fullName,
    planId,
    memberId,
    checkpointVisits,
    relationships,
    relationship,
    policy: { plans },
  } = memberProfile;
  summaryList.push({
    displayName: getDisplayName(intl, fullName, plans[planId]),
    planId,
    memberId,
    checkpointVisits,
    relationship,
  });
  if (!relationships) return summaryList;
  relationships.forEach(
    // eslint-disable-next-line no-shadow
    ({ fullName, planId, memberId, checkpointVisits, relationship, status }) =>
      status !== TERMINATED &&
      summaryList.push({
        displayName: getDisplayName(intl, fullName, plans[planId]),
        planId,
        memberId,
        checkpointVisits,
        relationship,
      }),
  );
  return summaryList;
};

export const getPlanName = (memberId, consumerBenefits) => {
  const plan = consumerBenefits.find(b => b.memberId === memberId);

  return plan || { name: '' };
};

const getBenefitMemberDisplay = (intl, memberName, category) => {
  if (!category) return memberName;
  const [, name] = category.split(' ');

  return getDisplayName(intl, memberName, { name });
};

export const buildBenefitMemberList = (intl, memberProfile) => {
  const memberList = [];
  const { fullName, memberId, category, relationships } = memberProfile;

  memberList.push({
    displayName: getBenefitMemberDisplay(intl, fullName, category),
    memberId,
  });

  if (!relationships) return memberList;

  relationships.forEach(
    // eslint-disable-next-line no-shadow
    ({ fullName, category, status, memberId }) =>
      status !== TERMINATED &&
      memberList.push({
        displayName: getBenefitMemberDisplay(intl, fullName, category),
        memberId,
      }),
  );

  return memberList;
};

export const benefitNameToKey = name => name.replace(/ /g, '');

export const transformBenefitPlans = benefitPlans => {
  const plans = {};

  Object.keys(benefitPlans).forEach(planId => {
    const products = {};
    benefitPlans[planId].products.forEach(product => {
      const services = [];
      product.services.forEach(service => {
        service.details.forEach(detail => {
          services.push({
            id: service.id,
            name: service.name,
            metaText: service.metaText,
            ...detail,
          });
        });
      });
      products[product.productType] = {
        name: product.name,
        footnote: product.footnote,
        panelLabel: product.panelLabel,
        nonPanelLabel: product.nonPanelLabel,
        freeChoiceLabel: product.freeChoiceLabel,
        unlimitedCheckpoint: product.unlimitedCheckpoint,
        services,
      };
    });
    plans[planId] = products;
  });

  return plans;
};

const createCheckpointsMap = checkpoints => {
  if (!checkpoints) return {};
  const checkpointsMap = {};
  checkpoints.forEach(({ serviceId, usedCount }) => {
    checkpointsMap[serviceId] = usedCount;
  });
  return checkpointsMap;
};

export const transformSelectedPlanWithMemberDetails = (
  intl,
  checkpointVisitsData,
  selectedPlanDetails,
  availableWalletBalance,
  relationship,
) => {
  const checkpointsMap = createCheckpointsMap(checkpointVisitsData);
  if (!selectedPlanDetails) return {};
  Object.entries(selectedPlanDetails).forEach(([productType, plan]) => {
    plan.services = plan.services.filter(
      ({ forRelationship }) =>
        productType !== WELLNESS_FLEXIBLE_SPENDING ||
        relationship === forRelationship,
    );
    plan.services.forEach(details => {
      if (details.checkpointVisits) {
        details.checkpointVisits.balance =
          +details.checkpointVisits.limit - (+checkpointsMap[details.id] || 0);
      }
      if (productType === WELLNESS_FLEXIBLE_SPENDING) {
        const maxWalletBalance = details.annualLimitText; // formatAmount(intl, details.annualLimitText);

        let messageKey = 'me.tabs.benefits.label.wellness.nonPanelDoctorValue';
        let defaultMessage = `${availableWalletBalance} out of ${maxWalletBalance} left`;
        if (
          availableWalletBalance === undefined ||
          availableWalletBalance === null
        ) {
          messageKey =
            'me.tabs.benefits.label.wellness.nonPanelDoctorValue.noBalance';
          defaultMessage = `- out of ${maxWalletBalance} left`;
        }
        details.nonPanelVisit = formatMessage(
          intl,
          messageKey,
          defaultMessage,
          {
            availableWalletBalance,
            maxWalletBalance,
          },
        );
      }
    });
  });
  return selectedPlanDetails;
};

export const transformLifeBenefits = groupLifeBenefits =>
  groupLifeBenefits.map(
    ({ benefitName, descriptionLabel, sumAssuredLabel, benefitDetails }) => {
      return {
        fullWidth: true,
        benefitName,
        panelLabel: descriptionLabel,
        nonPanelLabel: sumAssuredLabel,
        subLimits: [],
        annualLimit: {},
        benefitDetails: benefitDetails.map(detail => ({
          ...detail,
          isLifeBenefit: true,
          subBenefits: [
            {
              panel: detail.description,
              nonPanel: detail.sumAssured,
            },
          ],
        })),
      };
    },
  );

export const transformSublimit = subLimits =>
  subLimits
    .filter(({ text }) => !!text)
    .map(({ text, label, footnoteSuperscript, footnote, ...rest }) => ({
      benefitType: label,
      footnoteSuperscript,
      footnote,
      subBenefits: [{ ...rest }],
      checkpoint: text,
    }));
