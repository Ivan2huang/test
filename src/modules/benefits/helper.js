import { IsDependent, IsEmployee } from '../../helpers/roles';
import { IsCategorySpouse, IsCategoryChild } from '../../helpers/relationships';

const getCoPayment = (outPatientServices, serviceId) => {
  const { coPayment } = outPatientServices.find(
    service => service.id === serviceId,
  ).details[0];

  return {
    coPayment,
    coPaymentText: `$${coPayment}`,
  };
};

const onlyUnique = (value, index, self) => {
  return self.indexOf(value) === index;
};

const getUniquePlanIds = benefits => {
  const {
    member: { planId },
    relationships,
  } = benefits;

  let uniquePlanIdList = [];
  if (relationships.length !== 0) {
    const relationshipPlanId = relationships.map(r => r.planId);
    uniquePlanIdList = relationshipPlanId.filter(onlyUnique);
  }

  const allPlanIds = [planId, ...uniquePlanIdList];
  return allPlanIds.filter(onlyUnique);
};

const transformAndFilterPolicyPayloadByMembersBenefits = (
  policyDetails,
  benefits,
) => {
  let updatedPlans;
  if (policyDetails) {
    const uniquePlanIds = getUniquePlanIds(benefits);
    const { plans } = policyDetails;
    updatedPlans = plans.reduce((acc, plan) => {
      const { id, ...rest } = plan;
      if (uniquePlanIds.includes(id)) {
        acc[plan.id] = rest;
      }
      return acc;
    }, {});
  }

  return {
    ...policyDetails,
    plans: updatedPlans,
  };
};

const getCoPaymentDetailsByPlanId = (plans, planId) => {
  const outPatientServices = plans
    .find(plan => plan.id === planId)
    .products.find(product => product.ehealthcard === true).services;
  const gp = getCoPayment(outPatientServices, 'GP');
  const sp = getCoPayment(outPatientServices, 'SP');
  const phy = getCoPayment(outPatientServices, 'PHY');
  return {
    GP: gp.coPayment,
    GPText: gp.coPaymentText,
    SP: sp.coPayment,
    SPText: sp.coPaymentText,
    PHY: phy.coPayment,
    PHYText: phy.coPaymentText,
  };
};

const getMembersCoPayments = (benefits, policy) => {
  let coPayments;
  if (benefits && policy) {
    const {
      member: { planId },
      relationships,
    } = benefits;
    const { plans } = policy;
    let uniquePlanIdList = [];
    if (relationships.length !== 0) {
      const relationshipPlanId = relationships.map(r => r.planId);
      uniquePlanIdList = relationshipPlanId.filter(onlyUnique);
    }
    const allPlanIds = [planId, ...uniquePlanIdList];
    const uniquePlanIds = allPlanIds.filter(onlyUnique);

    coPayments = uniquePlanIds.reduce((acc, id) => {
      const result = getCoPaymentDetailsByPlanId(plans, id);
      acc[id] = result;
      return acc;
    }, {});
  }

  return coPayments;
};

export const orderMembersByRole = members => {
  const getPriorityOfRole = (role, relationshipCategory) => {
    if (IsEmployee(role)) return 1;
    if (IsDependent(role) && IsCategorySpouse(relationshipCategory)) return 2;
    if (IsDependent(role) && IsCategoryChild(relationshipCategory)) return 3;
    return 4;
  };

  const compareByRole = (firstMember, secondMember) => {
    const first = getPriorityOfRole(
      firstMember.role,
      firstMember.relationshipCategory,
    );
    const second = getPriorityOfRole(
      secondMember.role,
      secondMember.relationshipCategory,
    );
    return first - second;
  };

  return members.sort(compareByRole);
};

export const enrichUserProfileWithMemberProfileOrder = userProfile => {
  const { relationships, ...rest } = userProfile;
  const outerProfile = {
    fullName: rest.fullName,
    memberId: rest.memberId,
    membershipNumber: rest.membershipNumber,
    relationship: rest.relationship,
    relationshipToEmployee: rest.relationshipToEmployee,
    relationshipCategory: rest.relationshipCategory,
    role: rest.role,
    email: rest.email,
    workEmail: rest.workEmail,
    hasLoggedIn: rest.hasLoggedIn,
    policyNumber: rest.policy ? rest.policy.policyNumber : '',
    certificateNumber: rest.certificateNumber,
    profilePolicyNumber: rest.profilePolicyNumber || '',
    profileCertificateNumber: rest.profileCertificateNumber || '',
    contactNumber: rest.contactNumber,
  };

  const members = [{ ...outerProfile }, ...relationships];
  const memberProfileOrder = orderMembersByRole(members);

  return { ...userProfile, memberProfileOrder };
};

export const enrichProfileWithBenefits = (
  memberProfile,
  benefits,
  healthCards,
  policy,
) => {
  if (!memberProfile) {
    return undefined;
  }

  let result = {};

  if (memberProfile && benefits) {
    const {
      fullName,
      dependants,
      preferredLocale,
      isEdmOptedOut,
      isTermsAccepted,
      role,
      relationship,
      relationshipToEmployee,
      relationshipCategory,
      email,
      workEmail,
      hasLoggedIn,
      contactNumber,
    } = memberProfile;
    const { member, relationships } = benefits;

    const memberRelationships = dependants.map(d => {
      const dependantInfo =
        relationships.find(r => r.memberId === d.memberId) || {};
      const { planId, checkpointVisits, membershipNumber } = dependantInfo;

      return {
        ...d,
        planId,
        checkpointVisits,
        membershipNumber,
      };
    });

    result = {
      fullName,
      preferredLocale,
      isEdmOptedOut,
      isTermsAccepted,
      email,
      workEmail,
      role,
      relationship,
      relationshipToEmployee,
      relationshipCategory,
      hasLoggedIn,
      ...member,
      profilePolicyNumber: memberProfile.policyNumber,
      profileCertificateNumber: memberProfile.certificateNumber,
      relationships: memberRelationships,
      coPayments: getMembersCoPayments(benefits, policy),
      healthCards,
      policy: transformAndFilterPolicyPayloadByMembersBenefits(
        policy,
        benefits,
      ),
      contactNumber,
    };
  } else {
    result = {
      ...memberProfile,
      relationships: memberProfile.dependants,
      email: memberProfile.email,
      workEmail: memberProfile.workEmail,
      dependants: [],
    };
  }

  return result;
};
