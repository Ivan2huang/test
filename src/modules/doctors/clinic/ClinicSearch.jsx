import React, { useEffect, useRef, useState } from 'react';
import * as PropTypes from 'prop-types';
import { injectIntl } from 'react-intl';
import { compose } from 'redux';

import {
  Box,
  ClickAwayListener,
  IconButton,
  InputAdornment,
  TextField,
  withStyles,
} from '@material-ui/core';

import { formatMessage } from '../../../helpers/helpers';
import ClinicSearchSuggestions from './ClinicSearchSuggestions';
import { getRecentSearch, saveRecentSearch } from './helper';
import { logAction } from '../../../helpers/firebase';
import { CATEGORIES } from '../../../constants/analytics';
import Search from '../../../icons/Search';

const Styles = () => ({
  searchField: {
    margin: 0,
  },
});

const ClinicSearch = ({
  intl,
  classes,
  areas,
  searchClinic,
  resetMap,
  open,
  setOpen,
}) => {
  const [recentSearches, setRecentSearches] = useState([]);
  const [searchText, setSearchText] = useState('');
  const searchKeywordRef = useRef(null);
  useEffect(() => {
    setRecentSearches(getRecentSearch());
  }, []);

  const handleSearch = (text, searchBy) => {
    const trimmedSearchText = text.trim();
    searchClinic(trimmedSearchText, searchBy);
    if (trimmedSearchText) {
      saveRecentSearch(trimmedSearchText, searchBy);
      setRecentSearches(getRecentSearch());
      logAction({
        category: CATEGORIES.PANEL_CLINIC_SEARCH,
        action: 'Search',
      });
    }
    setSearchText(trimmedSearchText);
    searchKeywordRef.current.blur();
    setOpen(false);
  };

  const showAllClinics = () => {
    setOpen(false);
    resetMap();
    setSearchText('');
  };

  return (
    <ClickAwayListener onClickAway={() => setOpen(false)}>
      <Box position="relative">
        <TextField
          fullWidth
          variant="outlined"
          data-testid="search-bar-field"
          autoComplete="off"
          inputRef={searchKeywordRef}
          value={searchText}
          onChange={e => setSearchText(e.target.value)}
          classes={{ root: classes.searchField }}
          label={formatMessage(
            intl,
            'clinic.searchClinics.label.searchField',
            'Search clinic or location',
          )}
          onKeyPress={event => {
            if (event.key === 'Enter') {
              handleSearch(searchText);
            }
          }}
          InputProps={{
            endAdornment: (
              <InputAdornment position="end">
                <IconButton
                  data-testid="btn-clinic-search"
                  aria-label="search"
                  onClick={() => handleSearch(searchText)}
                >
                  <Search data-testid="img-search" />
                </IconButton>
              </InputAdornment>
            ),
          }}
          onFocus={() => setOpen(true)}
        />
        <ClinicSearchSuggestions
          open={open}
          areas={areas}
          onSearch={handleSearch}
          recentSearches={recentSearches}
          showAllClinics={showAllClinics}
        />
      </Box>
    </ClickAwayListener>
  );
};

ClinicSearch.propTypes = {
  intl: PropTypes.shape({}).isRequired,
  classes: PropTypes.shape({
    searchField: PropTypes.string.isRequired,
  }).isRequired,
  areas: PropTypes.shape({}).isRequired,
  searchClinic: PropTypes.func.isRequired,
  resetMap: PropTypes.func.isRequired,
  open: PropTypes.bool.isRequired,
  setOpen: PropTypes.func.isRequired,
};

export default compose(
  injectIntl,
  withStyles(Styles),
)(ClinicSearch);
