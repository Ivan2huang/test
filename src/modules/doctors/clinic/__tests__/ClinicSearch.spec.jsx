/* eslint-disable react/prop-types,react/button-has-type */
import React from 'react';
import { render, fireEvent, act } from '@testing-library/react';

import ClinicSearch from '../ClinicSearch';
import withIntl from '../../../../i18n/withIntlProvider';
import withTheme from '../../../../themes/withThemeProvider';
import { getRecentSearch, saveRecentSearch } from '../helper';
import { logAction } from '../../../../helpers/firebase';

jest.mock('../../../../helpers/helpers', () => ({
  formatMessage: jest.fn((intl, id, defaultMessage) => defaultMessage),
}));

jest.mock('../helper', () => ({
  getRecentSearch: jest.fn(),
  saveRecentSearch: jest.fn(),
}));

jest.mock('../../../../helpers/firebase', () => ({
  logAction: jest.fn(),
}));

jest.mock(
  '../../../../uiComponents/Typography',
  () => ({ children, ...rest }) => <div {...rest}>{children}</div>,
);

jest.mock(
  '../ClinicSearchSuggestions',
  () => ({ open, areas, onSearch, children, showAllClinics }) => (
    <>
      <div data-testid="popup-suggestions" data-open={open} data-areas={areas}>
        Clinic Search Suggestions
      </div>
      <button data-testid="show-all-clinics" onClick={() => showAllClinics()}>
        Show all clinics
      </button>
      <button
        data-testid="popup-suggestion-item"
        onClick={() => onSearch('Central1', 'district')}
      >
        Suggestion
      </button>
      {children}
    </>
  ),
);

describe('ClinicSearch Component', () => {
  let props;

  const setUp = (componentProps = props) => {
    const Component = withIntl(withTheme(ClinicSearch));
    return render(<Component {...componentProps} />);
  };

  beforeEach(() => {
    props = {
      searchClinic: jest.fn(),
      areas: {
        area1: ['Central1', 'Central3'],
        area2: ['Central2'],
      },
      resetMap: jest.fn(),
      open: false,
      setOpen: jest.fn(),
    };
  });

  afterEach(() => {
    jest.clearAllMocks();
    logAction.mockClear();
  });

  it('should match snapshot', () => {
    const { container } = setUp();

    expect(container).toMatchSnapshot();
  });

  it('should search clinics and save recent searches with search keyword when clicked on search', () => {
    const { getByTestId } = setUp();

    const searchField = getByTestId('search-bar-field').querySelector('input');
    const searchButton = getByTestId('btn-clinic-search');

    const searchText = 'test';
    fireEvent.change(searchField, { target: { value: searchText } });

    fireEvent.click(searchButton);

    expect(props.searchClinic).toHaveBeenCalledTimes(1);
    expect(props.searchClinic).toHaveBeenCalledWith(searchText, undefined);
    expect(getRecentSearch).toHaveBeenCalledTimes(2);
    expect(saveRecentSearch).toHaveBeenCalledWith('test', undefined);
  });

  it('should handle search when search text is empty', () => {
    const { getByTestId } = setUp();

    const searchField = getByTestId('search-bar-field').querySelector('input');
    const searchButton = getByTestId('btn-clinic-search');

    const searchText = '   ';
    fireEvent.change(searchField, { target: { value: searchText } });

    fireEvent.click(searchButton);

    expect(props.searchClinic).toHaveBeenCalledTimes(1);
    expect(props.searchClinic).toHaveBeenCalledWith(
      searchText.trim(),
      undefined,
    );
    expect(getRecentSearch).toHaveBeenCalledTimes(1);
    expect(saveRecentSearch).not.toHaveBeenCalled();
  });

  it('should handle search when search contain empty spaces', () => {
    const { getByTestId } = setUp();

    const searchField = getByTestId('search-bar-field').querySelector('input');
    const searchButton = getByTestId('btn-clinic-search');

    const searchText = '   test  ';
    fireEvent.change(searchField, { target: { value: searchText } });

    fireEvent.click(searchButton);

    expect(props.searchClinic).toHaveBeenCalledTimes(1);
    expect(props.searchClinic).toHaveBeenCalledWith(
      searchText.trim(),
      undefined,
    );
    expect(getRecentSearch).toHaveBeenCalledTimes(2);
    expect(saveRecentSearch).toHaveBeenCalledWith('test', undefined);
  });

  it('should call searchClinic with search keyword when pressed enter key', () => {
    const { getByTestId } = setUp();

    const searchField = getByTestId('search-bar-field').querySelector('input');

    const searchText = 'test';
    fireEvent.change(searchField, { target: { value: searchText } });

    fireEvent.keyPress(searchField, {
      key: 'Enter',
      code: 13,
      charCode: 13,
    });

    expect(props.searchClinic).toHaveBeenCalledTimes(1);
    expect(props.searchClinic).toHaveBeenCalledWith(searchText, undefined);
  });

  it('should not call searchClinic with search keyword when pressed any key other than enter key', () => {
    const { getByTestId } = setUp();

    const searchField = getByTestId('search-bar-field').querySelector('input');

    const searchText = 'test';
    fireEvent.change(searchField, { target: { value: searchText } });

    fireEvent.keyPress(searchField, {
      key: 'a',
      code: 97,
      charCode: 97,
    });

    expect(props.searchClinic).toHaveBeenCalledTimes(0);
  });

  it('should open suggestion popup on input focus', () => {
    const { getByTestId } = setUp();

    const searchField = getByTestId('search-bar-field').querySelector('input');
    fireEvent.focus(searchField);

    expect(props.setOpen).toHaveBeenCalledWith(true);
  });

  it('should call searchClinic and close suggestion pop up on any suggestion click', () => {
    const { getByTestId } = setUp();

    const searchField = getByTestId('search-bar-field').querySelector('input');
    const button = getByTestId('popup-suggestion-item');
    const popper = getByTestId('popup-suggestions');
    fireEvent.click(button);
    const open = popper.getAttribute('data-open');

    expect(open).toBe('false');
    expect(searchField.value).toBe('Central1');
    expect(props.searchClinic).toHaveBeenCalledTimes(1);
    expect(props.searchClinic).toHaveBeenCalledWith('Central1', 'district');
  });

  it('should close suggestion pop up when clicked outside', () => {
    const { getByTestId } = setUp();

    const searchField = getByTestId('search-bar-field').querySelector('input');
    fireEvent.focus(searchField);
    const popper = getByTestId('popup-suggestions');
    act(() => {
      document.querySelector('body').click();
    });
    const open = popper.getAttribute('data-open');

    expect(open).toBe('false');
  });

  it('should loose search field focus after keying text and on press enter', () => {
    const { getByTestId } = setUp();

    const searchField = getByTestId('search-bar-field').querySelector('input');

    const searchText = 'test';
    fireEvent.change(searchField, { target: { value: searchText } });
    fireEvent.keyPress(searchField, {
      key: 'Enter',
      code: 13,
      charCode: 13,
    });

    expect(document.activeElement).not.toEqual(searchField);
  });

  it('should handle show all clinics when show all clinics clicked', () => {
    const { getByTestId } = setUp();

    const searchField = getByTestId('search-bar-field').querySelector('input');
    const button = getByTestId('show-all-clinics');
    const popper = getByTestId('popup-suggestions');
    fireEvent.click(button);
    const open = popper.getAttribute('data-open');

    expect(open).toBe('false');
    expect(searchField.value).toBe('');
    expect(props.resetMap).toHaveBeenCalledTimes(1);
  });

  it('should not call firebase when search change without empty value', () => {
    const { getByTestId } = setUp();

    fireEvent.change(getByTestId('search-bar-field').querySelector('input'), {
      target: {
        value: '',
      },
    });

    expect(logAction).not.toHaveBeenCalled();
  });
});
