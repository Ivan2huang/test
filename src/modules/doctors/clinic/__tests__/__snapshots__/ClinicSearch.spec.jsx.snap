// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClinicSearch Component should match snapshot 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-3"
  >
    <div
      class="MuiFormControl-root MuiTextField-root ClinicSearch-searchField-1 MuiFormControl-fullWidth"
      data-testid="search-bar-field"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
        data-shrink="false"
      >
        Search clinic or location
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
      >
        <input
          aria-invalid="false"
          autocomplete="off"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
          type="text"
          value=""
        />
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
        >
          <button
            aria-label="search"
            class="MuiButtonBase-root MuiIconButton-root"
            data-testid="btn-clinic-search"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                data-testid="img-search"
                focusable="false"
                role="presentation"
                style="width: 1.125rem; height: 1.125rem;"
                viewBox="0 0 24 24"
              >
                <path
                  d="M2.92924 17.0707C4.88124 19.024 7.44124 20 9.99991 20C12.3092 20 14.6132 19.1973 16.4772 17.608L22.8346 23.964L23.9666 22.8333L17.6092 16.476C20.9599 12.5467 20.7866 6.64133 17.0732 2.928C15.1186 0.976 12.5586 0 9.99991 0C7.44124 0 4.88124 0.976 2.92924 2.92933C-0.976089 6.83467 -0.976089 13.1653 2.92924 17.0707ZM1.59991 10C1.59991 7.756 2.47458 5.64667 4.05991 4.06C5.64524 2.47333 7.75591 1.6 9.99991 1.6C12.2439 1.6 14.3532 2.47333 15.9399 4.06C19.2146 7.33467 19.2146 12.664 15.9399 15.94C14.3532 17.5267 12.2439 18.4 9.99991 18.4C7.75591 18.4 5.64658 17.5267 4.05991 15.94C2.47324 14.3533 1.59991 12.244 1.59991 10Z"
                  fill="#333333"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-91 MuiOutlinedInput-notchedOutline"
          style="padding-left: 8px;"
        >
          <legend
            class="PrivateNotchedOutline-legend-92"
            style="width: 0.01px;"
          >
            <span>
              â€‹
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      data-areas="[object Object]"
      data-open="false"
      data-testid="popup-suggestions"
    >
      Clinic Search Suggestions
    </div>
    <button
      data-testid="show-all-clinics"
    >
      Show all clinics
    </button>
    <button
      data-testid="popup-suggestion-item"
    >
      Suggestion
    </button>
  </div>
</div>
`;
